### PR TITLE
Add new RUN command as a start of all deploy scripts

### PR DESCRIPTION
**Pull request by azizova-leyla**
### Description

New pattern of calling deploy scripts:
1. Call cloud/shared/run.py --command=[setup, deploy, destroy] --tag=IMAGE_TAG --config=CONFIG_FILE
2. Inside run.py load config, validate variables and then call out to actual command.

This centralize config loading to one file, and will be used in the dockerized version.

I'm planning to delete bash versions of scripts from cloud/shared/bin later, leaving for now so all current scripts work.

Also fixed actions that were not setting --tag properly.

## Release notes:

Describe the change as it should be communicated to partners in the release notes.

### Checklist

- [ ] Added the correct label: < feature | bug | dependencies | infrastructure | ignore-for-release >
- [ ] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### Issue(s)

First step to [#3222](https://github.com/civiform/civiform/issues/3222)

### CODE REVIEW

File: cloud/shared/bin/deploy.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import subprocess
+import importlib
+import os
+from cloud.shared.bin.lib.config_loader import ConfigLoader
```
Comment (bion) [inline cloud/shared/bin/deploy.py:4]: nit: `ConfigLoader` isn't used in this file

File: cloud/shared/bin/destroy.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
+#! /usr/bin/env python3
+
+from cloud.shared.bin.lib.config_loader import ConfigLoader
```
Comment (bion) [inline cloud/shared/bin/destroy.py:3]: nit: `ConfigLoader` isn't used in this file

File: cloud/shared/bin/lib/config_loader.py [hunk 3]
```diff
@@ -45,20 +47,35 @@ def use_local_backend(self):
...
+            exit(f'Cannot find file {full_config_path}')
+        print(f'Getting config from {full_config_path}')
+        command = shlex.split(
+            f'env -i bash -c "source {full_config_path} && env"')
+        proc = subprocess.Popen(command, stdout=subprocess.PIPE)
+        for line in proc.stdout:
+            (key, _, value) = line.decode().partition("=")
+            os.environ[key] = value.strip()
+        proc.communicate()
+
```
Comment (bion) [inline cloud/shared/bin/lib/config_loader.py:67]: suggestion (blocking): this could use a comment to describe what's happening, and potentially extracted to a method with a more descriptive method name

File: cloud/shared/bin/lib/config_loader.py [hunk 4]
```diff
@@ -124,7 +139,10 @@ def get_cloud_provider(self):
         return self.configs.get('CIVIFORM_CLOUD_PROVIDER')
 
     def get_template_dir(self):
-        return self.configs.get('TERRAFORM_TEMPLATE_DIR')
+        template_dir = self.configs.get('TERRAFORM_TEMPLATE_DIR')
+        if template_dir is None or not os.path.exists(template_dir):
+            exit(f'Cound not find template directory {template_dir}')
```
Comment (bion) [inline cloud/shared/bin/lib/config_loader.py:144]:
```suggestion
            exit(f'Could not find template directory {template_dir}')
```

File: cloud/shared/bin/run.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        '--tag', help='Civiform image tag. Required for Setup and Deploy.')
+    parser.add_argument(
+        '--command',
+        help='Command to run. If ommited will validate config and exit.')
+    parser.add_argument(
+        '--config',
+        default='civiform_config.sh',
+        help='Path to civiform config file.')
```
Comment (bion) [inline cloud/shared/bin/run.py:24]: nit: to disambiguate with `application.conf` maybe say "Path to CiviForm deployment config file"

Comment (bion) [inline cloud/shared/bin/run.py:24]: nit: to disambiguate with `application.conf` maybe say "Path to CiviForm deployment config file"

**Review by bion — CHANGES_REQUESTED** [review]

**Review by bion — APPROVED** [review]

File: cloud/shared/bin/destroy.py [hunk 6]
```diff
@@ -0,0 +1,14 @@
+#! /usr/bin/env python3
+
+from cloud.shared.bin.lib import terraform
+from cloud.shared.bin.lib.setup_class_loader import get_config_specific_destroy
+"""
+Destroy.py destroys Civiform deployment.
+"""
```
Comment (nbeloglazov) [inline cloud/shared/bin/destroy.py:7]: Module description should be at the beginning of the file
https://google.github.io/styleguide/pyguide.html#382-modules

File: cloud/shared/bin/destroy.py [hunk 7]
```diff
@@ -0,0 +1,14 @@
+#! /usr/bin/env python3
```
Comment (nbeloglazov) [inline cloud/shared/bin/destroy.py:1]: You don't need shebang here given that this file is not main (and won't be run from command line)
https://google.github.io/styleguide/pyguide.html#37-shebang-line

File: cloud/shared/bin/deploy.py [hunk 8]
```diff
@@ -0,0 +1,16 @@
+import subprocess
+import importlib
+import os
+
+
+def run(config):
```
Comment (nbeloglazov) [inline cloud/shared/bin/deploy.py:6]: Can you add type hint here:

```
def run(config: ConfigLoader): 
   ...
```

This `run` function is not directly called anywhere so it might be challenging to figure out what `config` here is. 

Same for all other `run` functions in this PR.

File: cloud/shared/bin/lib/config_loader.py [hunk 9]
```diff
@@ -45,20 +47,37 @@ def use_local_backend(self):
...
 
-    def _load_config(self):
+    def _get_config_values_from_sh_file(self, config_file):
+        ## 1. Export all variables from the config into clean environment
+        ## 2. Set values in current environment
+        if not os.path.exists(config_file):
+            exit(f'Cannot find file {config_file}')
+        print(f'Getting config from {config_file}')
+        command = shlex.split(f'env -i bash -c "source {config_file} && env"')
+        proc = subprocess.Popen(command, stdout=subprocess.PIPE)
```
Comment (nbeloglazov) [inline cloud/shared/bin/lib/config_loader.py:61]: Interesting trick of reading shell variables without needing to pre-source it.

File: cloud/shared/bin/lib/config_loader.py [hunk 10]
```diff
@@ -45,20 +47,37 @@ def use_local_backend(self):
...
+        for line in proc.stdout:
+            (key, _, value) = line.decode().partition("=")
+            os.environ[key] = value.strip()
```
Comment (nbeloglazov) [inline cloud/shared/bin/lib/config_loader.py:64]: Hm. Do we actually need to expose these variables back to environment? Or can pass around as config?

File: cloud/shared/bin/run.py [hunk 11]
```diff
@@ -0,0 +1,63 @@
+#! /usr/bin/env python3
+
+import argparse
+import os
+import sys
+import importlib
+
+sys.path.append(os.getcwd())
```
Comment (nbeloglazov) [inline cloud/shared/bin/run.py:8]: Why is this here between imports? Can you add comment explaining this unconventional call?

File: cloud/shared/bin/run.py [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+
+    args = parser.parse_args()
+    if args.tag:
+        os.environ['TF_VAR_image_tag'] = args.tag
+    elif args.command is not None and args.command != 'destroy':
+        exit('--tag is required')
+
+    os.environ['TF_VAR_FILENAME'] = "setup.auto.tfvars"
+    os.environ['BACKEND_VARS_FILENAME'] = 'backend_vars'
+    os.environ['TERRAFORM_PLAN_OUT_FILE'] = 'terraform_plan'
```
Comment (nbeloglazov) [inline cloud/shared/bin/run.py:34]: FWIW it's odd to see these variables here. Would be better to set them only when we call terraform, and maybe pass them just to the call itself, if it's possible, rather than setting global variables. But it's probably requires more work.

**Review by nbeloglazov — CHANGES_REQUESTED** [review]

File: cloud/shared/bin/lib/config_loader.py (hunk 10 above)
Comment (azizova-leyla) [reply cloud/shared/bin/lib/config_loader.py:64]: Currently ConfigLoader reads some values from env variables, and I don't want to change that in this PR.
Next big change is to make config yaml and all of this code will be changed. At that point ConfigLoader will be one place to get all the values.
It will likely require rewrite of some bash scripts that we still have.

**Review by nbeloglazov — APPROVED** [review]
